### PR TITLE
Add default case rule

### PR DESCRIPTION
**Pull request by rakaramos**
Implements: https://github.com/realm/SwiftLint/issues/131

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/Idiomatic/DefaultCaseRule.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+public struct DefaultCaseRule: ASTRule, ConfigurationProviderRule, OptInRule, AutomaticTestableRule {
+
+    public var configuration = SeverityConfiguration(.warning)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "default_case",
+        name: "Default case",
+        description: "`default` statement on `switch` statement should be avoided.",
```
Comment (gobetti) [inline Source/SwiftLintFramework/Rules/Idiomatic/DefaultCaseRule.swift]: perhaps you meant "`default` *case* on `switch` statement should be avoided"?

**Comment by marcelofabri** [comment]
This is not so useful right now because it'll trigger even when you need a `default` (switching on a non-enum value).

Maybe we could use the new analyze mode to get more information about the expression being switched on and check if it's an enum?

**Comment by rakaramos** [comment]
Since [0192-non-exhaustive-enums](https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md) was implemented on Swift 5, I think  this rule will be useless (unless for people that don't adopt Swift 5 right away).

@marcelofabri do you think it still worth it?

**Comment by gobetti** [comment]
@rakaramos Swift 5 is going to introduce a distinction between frozen and non-frozen enums, which is not going to change anything for the currently already exhaustive enums, which are what makes your rule useful.

In other words, Swift 5 will not require you to _avoid the default case_, which is what your rule does. They will oblige you to *use* it, _when_ the enum is non-frozen. And for that use specifically, it will be `@unknown default`, which again is different from your rule (as long as your regex matches `default` but doesn't match the @unknown version).

I agree with the other comments: for this rule to be more useful than a pain, the type being switched on must be checked, it only makes sense where exhaustiveness exists, i.e. enums, booleans, perhaps something else. I don't think it will be an easy job, though, when tuples come to mind...

**Comment by Dschee** [comment]
Can we perhaps use [SwiftSyntax](https://github.com/apple/swift-syntax) to detect if the variable is of `enum` type? Does SwiftSyntax provide this info? I'm not really sure but kind of doubt it.

I also think this rule only should be merged if it can detect many common cases where actually an enum is used without producing any false positives. Do you still think it's useless after Swift 5 is released @rakaramos?

**Comment by rakaramos** [comment]
@dschee that's a good question! I think it will depend on how much checks the compiler is doing...I haven't played around with Swift 5, but would be nice to dig into the new `@unknown` keyword.

**Comment by marcelofabri** [comment]
> Does SwiftSyntax provide this info?

I don't think it does. We maybe can do it with our analyzer rules, but I don't know if it can handle variables with types that are declared in other files.

**Comment by JulianKahnert** [comment]
Hi @rakaramos @marcelofabri,
I would love to work on this rule.
What else needs to be done here?

**Pull request closed without merging**